### PR TITLE
Fix AMDGPU Function Calls Flag

### PR DESCRIPTION
**Pull request by aaronenyeshi**
Properly pass the flag into CompilerInvocation which needs explicit DriverArgs to CC1Args passing. This allows us to properly set up LangOpt AMDGPUFunctionCalls. Also change amdgpu-function-calls to a flag, -f[no-]amdgpu-function-calls . Let's set the default to false.

### CODE REVIEW

**Comment by aaronenyeshi** [comment]
I've verified that all tests passing by default. And then we still see the Function Call regressions when we change lit.cfg to pass in -fno-amdgpu-function-calls .

**Comment by aaronenyeshi** [comment]
Although this Flag is fixed, FC enabled still showing 154 failing HCC Unit tests on CTU, while FC disabled shows 0 failing.

File: lib/Driver/ToolChains/Hcc.cpp [hunk 1]
```diff
@@ -59,6 +61,11 @@ void HCCInstallationDetector::AddHCCIncludeArgs(const llvm::opt::ArgList &Driver
   if (IsValid) {
     CC1Args.push_back(DriverArgs.MakeArgString("-I" + IncPath + "/include"));
     CC1Args.push_back(DriverArgs.MakeArgString("-I" + IncPath + "/hcc/include"));
+
+    if (DriverArgs.hasFlag(options::OPT_famdgpu_function_calls,
+                           options::OPT_fno_amdgpu_function_calls,
+                           FunctionCallDefault))
+      CC1Args.push_back("-famdgpu-function-calls");
```
Comment (scchan) [inline lib/Driver/ToolChains/Hcc.cpp]: It doesn't seem like this is the right place to put this logic here since this is unrelated to include path processing.  Should this go into `construct_amdgpu_target_cmdargs` instead?

File: lib/Driver/ToolChains/Hcc.cpp [hunk 2]
```diff
@@ -91,8 +98,11 @@ void HCCInstallationDetector::AddHCCLibArgs(const llvm::opt::ArgList &Args, llvm
         CmdArgs.push_back(Args.MakeArgString(prefix + Lib));
     }
 
-    if (Args.hasArg(options::OPT_amdgpu_function_calls))
+    if (Args.hasFlag(options::OPT_famdgpu_function_calls,
+                     options::OPT_fno_amdgpu_function_calls,
+                     FunctionCallDefault)) {
       CmdArgs.push_back("--amdgpu-func-calls");
+    }
```
Comment (scchan) [inline lib/Driver/ToolChains/Hcc.cpp]: same comment as above

**Review by scchan — DISMISSED** [review]

File: lib/Driver/ToolChains/Hcc.cpp (hunk 1 above)
Comment (aaronenyeshi) [reply lib/Driver/ToolChains/Hcc.cpp]: I'm not sure if that function has access to the CC1 Args. That function only has Command arguments for Assembler and Linker, but this is a Frontend Option. The second part --amdgpu-func-calls can be moved to HCC:CXXAMPLINK::ConstructLinkerJob, but -famgpu-function-calls might need to go into Linux.cpp instead

Comment (aaronenyeshi) [reply lib/Driver/ToolChains/Hcc.cpp]: I've tested an alternative where I move --amdgpu-func-calls to `construct_amdgpu_target_cmdargs` but I move the -famdgpu-function-calls to `addClangTargetOptions`. another possibility is rename `addHCCIncludeArgs` to something generic like `addHCCCC1Args`

**Comment by aaronenyeshi** [comment]
Should we merge this so it makes it into HCC on next merge maintenance?

**Comment by aaronenyeshi** [comment]
Strangely, having `CC1Args.push_back("-famdgpu-function-calls");` in addHCCIncludeLibs has 154 failures, while having it in addClangTargetOptions has 156 failures. The difference is that the latter only sets Frontend flag for 2nd CompilerInvocation iteration, while first version sets it for both CompilerInvocations.

**Comment by aaronenyeshi** [comment]
The new location seems to have 154 errors again. 
  Expected Passes    : 542
  Expected Failures  : 35
  Unexpected Failures: 154

I will investigate these failures this week.
